pkg/scheduler/ugm/user_tracker.go

```diff
@@ -168,7 +168,7 @@ func (ut *UserTracker) canBeRemoved() bool {
 }
 
 func (ut *UserTracker) canRunApp(hierarchy []string, applicationID string) bool {
-	ut.Lock()
-	defer ut.Unlock()
+	ut.RLock()
+	defer ut.RUnlock()
 	return ut.queueTracker.canRunApp(hierarchy, applicationID, user)
```

Comment:
https://github.com/apache/yunikorn-core/blob/master/pkg/scheduler/ugm/queue_tracker.go#L101

it would change something internally, so is it ok to use read-lock? if using read lock is safe, should we update group_tacker also ? see https://github.com/apache/yunikorn-core/blob/master/pkg/scheduler/ugm/group_tracker.go#L149

Reply:
Ouch! Actually both `headroom()` and `canRunApp()` ends up modifying the map if the child tracker does not exist :( I was in a hurry to approve this change... @manirajv06 what do you think? As I can see it, `canRunApp()` is likely the first call which ends up creating the tracker hierarchy, unless there are limits. We need to go back to the Lock/Unlock pair.

Reply:
`headroom` has same issue. https://github.com/apache/yunikorn-core/blob/master/pkg/scheduler/ugm/queue_tracker.go#L210

Maybe we should either consider another naming or add some comments for `headroom` and `canRunApp`. It is error-prone to me.

Reply:
reverted.

Reply:
Logged jira for documenting lock behaviour. When I designed the tracker objects I explicitly kept the QueueTracker lock free.